test/e2e/utils.go

```diff
@@ -0,0 +1,231 @@
+// +build e2e
+
+package e2e
+
+import (
+	"fmt"
+	"os/exec"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// lookForStringInPodExec looks for expectedString in the output of command
+// executed in the specified pod container.
+func lookForStringInPodExec(ns, pod, container string, command []string, expectedString string, timeout time.Duration) error {
+	cmd, err := lookupCmdPath("oc")
+	if err != nil {
+		return err
+	}
+	args := []string{"exec", pod, "-c", container, fmt.Sprintf("--namespace=%v", ns), "--"}
+	args = append(args, command...)
+	if err := lookForString(cmd, args, expectedString, timeout); err != nil {
+		return err
+	}
+	return nil
+}
+
+// lookForStringInPodLog looks for the given string in the log of a specific pod container
+func lookForStringInPodLog(ns, pod, container, expectedString string, timeout time.Duration) error {
+	cmd, err := lookupCmdPath("oc")
+	if err != nil {
+		return err
+	}
+	args := []string{"logs", pod, "-c", container, fmt.Sprintf("--namespace=%v", ns)}
+	if err := lookForString(cmd, args, expectedString, timeout); err != nil {
+		return err
+	}
+	return nil
+}
+
+// lookupCmdPath looks-up the path of cmd returning the full command
+// path or an error.
+func lookupCmdPath(cmdPath string) (string, error) {
+	cmdPath, err := exec.LookPath(cmdPath)
+	if err != nil {
+		return "", fmt.Errorf("failed to find the %s executable in the directories named by the PATH environment variable.", cmdPath)
+	}
+	return cmdPath, nil
+}
+
+// lookForString looks for the given string using cmd and args repeatedly
+// until the timeout is reached or the string is found. Returns an error
+// if the string was not found.
+func lookForString(cmd string, args []string, expectedString string, timeout time.Duration) error {
+	for t := time.Now(); time.Since(t) < timeout; time.Sleep(2 * time.Second) {
+		result, err := runCmd(cmd, args)
+		if err != nil {
+			continue
+		}
+		if strings.Contains(result, expectedString) {
+			return nil
+		}
+	}
+	return fmt.Errorf("failed to find %q", expectedString)
+}
+
+// runCmd runs command cmd with arguments args and returns the output
+// of the command or an error.
+func runCmd(cmd string, args []string) (string, error) {
+	execCmd := exec.Command(cmd, args...)
+	result, err := execCmd.Output()
+	if err != nil {
+		return "", fmt.Errorf("failed to run command %q with args %q: %v", cmd, args, err)
+	}
+	return string(result), nil
+}
+
+// upstreamConfigMap returns a ConfigMap definition containing a
+// Corefile configured for the test upstream resolver.
+func upstreamConfigMap(name, ns string) *corev1.ConfigMap {
```

Comment:
This could be named more generically; e.g., `buildConfigMap`.

Reply:
ack